ambry-protocol/src/main/java/com.github.ambry.protocol/PutRequest.java

```diff
@@ -168,51 +171,80 @@ private int sizeExcludingBlobAndCrcSize() {
     return sizeExcludingBlobAndCrc;
   }
 
+  private void prepareBuffer() {
+    if (bufferToSend == null) {
+      // this is the first time this method was called, prepare the buffer to send the header and other metadata
+      // (everything except the blob content).
+      bufferToSend = ByteBuffer.allocate(sizeExcludingBlobAndCrcSize());
+      writeHeader();
+      int crcStart = bufferToSend.position();
+      bufferToSend.put(blobId.toBytes());
+      BlobPropertiesSerDe.serializeBlobProperties(bufferToSend, properties);
+      bufferToSend.putInt(usermetadata.capacity());
+      bufferToSend.put(usermetadata);
+      bufferToSend.putShort((short) blobType.ordinal());
+      short keyLength = blobEncryptionKey == null ? 0 : (short) blobEncryptionKey.remaining();
+      bufferToSend.putShort(keyLength);
+      if (keyLength > 0) {
+        bufferToSend.put(blobEncryptionKey);
+      }
+      bufferToSend.putLong(blobSize);
+      crc.update(bufferToSend.array(), bufferToSend.arrayOffset() + crcStart, bufferToSend.position() - crcStart);
+      nioBuffersFromBlob = blob.nioBuffers();
+      for (ByteBuffer bb : nioBuffersFromBlob) {
+        crc.update(bb);
+        // change it back to 0 since we are going to write it to the channel later.
+        bb.position(0);
+      }
+      crcBuf.putLong(crc.getValue());
+      crcBuf.flip();
+      bufferToSend.flip();
+    }
+  }
+
   @Override
   public long writeTo(WritableByteChannel channel) throws IOException {
     long written = 0;
-    if (sentBytes < sizeInBytes()) {
-      if (bufferToSend == null) {
-        // this is the first time this method was called, prepare the buffer to send the header and other metadata
-        // (everything except the blob content).
-        bufferToSend = ByteBuffer.allocate(sizeExcludingBlobAndCrcSize());
-        writeHeader();
-        int crcStart = bufferToSend.position();
-        bufferToSend.put(blobId.toBytes());
-        BlobPropertiesSerDe.serializeBlobProperties(bufferToSend, properties);
-        bufferToSend.putInt(usermetadata.capacity());
-        bufferToSend.put(usermetadata);
-        bufferToSend.putShort((short) blobType.ordinal());
-        short keyLength = blobEncryptionKey == null ? 0 : (short) blobEncryptionKey.remaining();
-        bufferToSend.putShort(keyLength);
-        if (keyLength > 0) {
-          bufferToSend.put(blobEncryptionKey);
+    try {
+      if (sentBytes < sizeInBytes()) {
+        prepareBuffer();
```

Comment:
optional:  I would suggest renaming this method and moving the "if" statement out of method. Something like this:
```
if (bufferToSend == null) {
 // this is the first time this method was called, prepare the buffer to send the header and other metadata
   prepareBufferToSend();
}
```

Reply:
sure